### PR TITLE
Feature/built value

### LINKED ISSUE #281: Dynamite: support ContentString and headers
#194 will remove/degrade the current Support for ContentStrings and headers.
This issue tracks this regression.

TL;DR It'll hopefully come back shortly but it's not as trivial as one might think.

Problem:
`built_value` works on Types instead of fields like `json_serializable`. This means that we can't add the ability to de serialize some fields differently (to built_value they are just strings).

We need to introduce a custom Type for ContentStrings (as done in the mentioned PR). We can't rely on custom logic in it's serializer (`_$ContentStringSerializer`) as this stage doesn't expose the data in a json format. This is especially needed when contentstrings are nested (`ContentString<...<SontentString<...>>>`

<details>
<summary>Why? - Small BuiltValue excursion :)</summary>
<br>
built_value will serialize the data to a custom Iterable<String> based format (`['$', 'TypeWireName', 'FieldWireName', 'serializedContent']`). We can and do apply the `StandardJsonPlugin` wich will take this format once and parse it from/to valid json. This brings improved performance and type support over just serializing to json.

Yes in theory this intermediate format could also then be parsed to xml (FRs exist) [Thinking of moving webdav to it once it exists]
</details>

How can we mitigate this?
We can register our own Plugin that will run before/after the StandardJsonPlugin. This way we can get raw access to the json/intermediate format and manipulate it as needed.

```dart
abstract class ContentString<T> implements Built<ContentString<T>, ContentStringBuilder<T>> {
  factory ContentString([final void Function(ContentStringBuilder<T>)? b]) = _$ContentString<T>;
  const ContentString._();

  /// decoded contentString
  T get content;

  static Serializer<ContentString<Object?>> get serializer => _$contentStringSerializer;
}

class ContentStringPlugin implements SerializerPlugin {
  @override
  Object? afterDeserialize(final Object? object, final FullType specifiedType) => object;

  @override
  Object? afterSerialize(final Object? object, final FullType specifiedType) {
    if (!specifiedType.root.toString().contains('ContentString')) {
      return object;
    }

    if (object is! Map<String, dynamic>) {
      throw ArgumentError('ContentStringPlugin can only be applied to Map<String, dynamic>. '
          'Please ensure the StandardJsonPlugin is applied and run before.');
    }

    final content = object['content'];

    return jsonEncode(content);
  }

  @override
  Object? beforeDeserialize(final Object? object, final FullType specifiedType) {
    if (!specifiedType.root.toString().contains('ContentString')) {
      return object;
    }

    if (object is! String) {
      throw ArgumentError('The serialized ContentString must be of type String. '
          'Please ensure the StandardJsonPlugin is applied and run before.');
    }

    final content = jsonDecode(object);

    return ['content', content];
  }

  @override
  Object? beforeSerialize(final Object? object, final FullType specifiedType) => object;
}
```

Note we throw errors when we detect invalid input indicating that the StandardJsonPlugin hasn't run yet. This can be due to forgetting to register it. Or wrong registration order (it needs to be registered first).

Cool the code is there why isn't it in dynamite?

built_value needs to have a `addBuilderFactory` registered fo every Type, Subtype combination (one for `BuiltList<String>`, one for `BuiltList<int>` ...). The generator will do this for us automatically for every combination it'll find in classes it generates for.
If we have a custom combination that doesn't occur otherwise (because it won't ever happen in a response but is needed for a request) we need to register those ourself. 

This can get ugly and I paused my work thinking about other possibilities before I fully commit to this route.

### PR DESCRIPTION
**Pull request by Leptopoda**
fixes #281  

what to expect:
- all tests passing :)

what not to expect:
- unrecognized_keys are always allowed => disallowUnrecognizedKeys is always false
- <s>custom serialization is untested and I am fairly sure this is not going to work as is</s>
- coverage is probably all over the place as I didn't bother checking or changing `// coverage:ignore-start`

---

<s>The last commit only patches the big generated spec which isn't that nice I know.</s>
Why did I change `ProvisioningApiAppInfoValue` from `oneOf` to `anyOf`?. As it contains both an Object and an Array (and a List is an Object) we basically have the issue that both are matching thus failing our assertion that only one should match. 

---

Feedback is welcome and I hope the multiple rebase work didn't break anything.
Marking as draft so the above issues can be ironed out.

### CODE REVIEW

**Comment by provokateurin** [comment]
The output looks really good! I wonder if it would be possible to continue using e.g. List instead of BuiltList, because we'd have to either export that from the package or require users to import it manually. Is there a good reason to not do it?

**Comment by Leptopoda** [comment]
Built_value wont run when we don't use other built objects. It's the same with built_map.
 
 So sadly no :(

**Comment by Leptopoda** [comment]
I consider this as ready for now. All tests are passing and to keep it small we should probably add the other stuff (like the stuff discussed in #193) in later PRs.

Also the next two to three weeks are full with exams so I don't have time to do anything further as of now.

**Comment by provokateurin** [comment]
I have a bunch of fixes for dynamite in a few branches. I think it would be a lot easier to merge them first and let you rebase. Would that be fine for you?

**Comment by Leptopoda** [comment]
I hate rebasing but who doesn't :)

but I agree that it's probably easier to rebase this than the other multiple small things xou have.

**Comment by provokateurin** [comment]
Changes are in: https://github.com/provokateurin/nextcloud-neon/pull/199

**Comment by Leptopoda** [comment]
So I can rebase now :eyes: ?

**Comment by provokateurin** [comment]
Yep

**Comment by Leptopoda** [comment]
I hope I didn't break anything with my rebase :sweat_smile:

**Comment by Leptopoda** [comment]
That's odd as the script should format it on generation 🤔

I'll reformat and push tomorrow

**Comment by provokateurin** [comment]
@Leptopoda Could you rebase? I hope I can finally review this PR soon

**Comment by Leptopoda** [comment]
I can rebase now but I haven't fixed the last one yet (contentStrings and extraJsonSerializableValues are not working rn)

I'll have a look into them tomorrow

**Comment by provokateurin** [comment]
No problem

**Comment by Leptopoda** [comment]
I'm not a fan of https://github.com/provokateurin/nextcloud-neon/pull/194/commits/fe7452634994e6c83e436890908492c1c69af551 and https://github.com/provokateurin/nextcloud-neon/pull/194/commits/7f0131dcf62832deab5bf4279659b71a6c49f953 as they add quite a bit of redundant code (especially regarding the serialization)

I left it as is for now and plan to unify quite a bit in the future.

**Comment by provokateurin** [comment]
Is it possible to split this into multiple PRs? I find it hard to review such a big change, especially because of moved and reformatted code which makes it really hard to spot differences. I'm really sorry that I haven't reviewed this yet :/

**Comment by Leptopoda** [comment]
I find it quite useful reviewing each commit on its own and also having the diff side by side instead of inline.

How would you prefer it?
maybe you can create a branch named built_value where I can make can make multiple PRs against?
Making multiple PRs against main is dumb as no commit will pass the tests on it's own.

when running the builder each commit should pass static analysis on it's own though :woman_shrugging:

**Comment by provokateurin** [comment]
I can try doing commit by commit, but could you submit changes like https://github.com/provokateurin/nextcloud-neon/pull/194/commits/9bd3b47d2837bfc6868e5704c9d163b7824ee945 separately?

**Comment by provokateurin** [comment]
Once you rebase I'll review. I think you need to re-order https://github.com/provokateurin/nextcloud-neon/pull/194/commits/b9347d541673ae773f563e777172ece8520ccd0d and https://github.com/provokateurin/nextcloud-neon/pull/194/commits/d001daba851712e4fe28812629a6572b57530e9a

**Comment by Leptopoda** [comment]
Ok I rebased and just noticed that we might/will need to relax our assertions for OneOf/AnyOf serialization.
Something like

```json
"data": {
  "oneOf": [
    {
      "type": "array",
      "items": {
        "type": "object"
      }
    },
    {
      "$ref": "#/components/schemas/PublicStatus"
    }
  ]
}
```
will generate

```dart
  @override
  NextcloudUserStatusGetStatus_Ocs_Data deserialize(
    Serializers serializers,
    Object data, {
    FullType specifiedType = FullType.unspecified,
  }) {
    final result = NextcloudUserStatusGetStatus_Ocs_DataBuilder()..data = JsonObject(data);
    try {
      result.builtListJsonObject = BuiltList<JsonObject>((data as List).map((final e) => JsonObject(e))).toBuilder();
    } catch (_) {}
    try {
      result.userStatusStatus = NextcloudUserStatusStatus.fromJson(data as Object).toBuilder();
    } catch (_) {}
    assert([result._builtListJsonObject, result._userStatusStatus].where((final x) => x != null).length == 1,
        'Need oneOf for ${result._data}');
    return result.build();
  }
```

BuiltValue will use List<Object> internally to represent the serialized data (JSON => List<Object> --Serializer--> Deserialized Object) this means that a serialized Object will almost every time match the type List<dynamic> thus failing the assertion as both values won't be null.

What's the best thing to do :thinking: ?

**Comment by provokateurin** [comment]
Yeah I also realized that. Maybe we should just drop all the asserts? We could also only drop the assert if we detect that one of the cases will always match another case (like `dynamic` for everything or `List<dynamic>` matching `List<String>`).

**Comment by provokateurin** [comment]
Yeah let's drop them for now (I don't think most or any generators uses them anyway).

**Comment by Leptopoda** [comment]
Ok I just relaxed the assert to allso multiple matches effectively making oneOf equivalent to anyOf.
There should still be at least one value present and we can assert that.

I'm running the test and will push

**Comment by provokateurin** [comment]
Can you submit the assert relaxation separately (since they actually affect existing code)?

**Comment by Leptopoda** [comment]
Ok the format bug again.

In favor of not wasting CI minutes I'll fix this with the next push

**Comment by provokateurin** [comment]
Just as an experiment I rebased this branch onto https://github.com/provokateurin/nextcloud-neon/tree/feature/openapi-automatic and got the following error message when trying to run dynamite:
```log
#0      TypeResultBase.decode (package:dynamite/src/type_result/base.dart:28:9)
#1      OpenAPIBuilder.build.<anonymous closure>.<anonymous closure> (package:dynamite/src/openapi_builder.dart:907:79)
#2      _$MethodBuilder.update (package:code_builder/src/specs/method.g.dart:333:33)
#3      new _$Method (package:code_builder/src/specs/method.g.dart:38:29)
#4      OpenAPIBuilder.build.<anonymous closure> (package:dynamite/src/openapi_builder.dart:616:25)
#5      _$ClassBuilder.update (package:code_builder/src/specs/class.g.dart:292:33)
#6      new _$Class (package:code_builder/src/specs/class.g.dart:34:28)
#7      OpenAPIBuilder.build (package:dynamite/src/openapi_builder.dart:411:11)
<asynchronous suspension>
#8      runBuilder.buildForInput (package:build/src/generate/run_builder.dart:52:7)
<asynchronous suspension>
#9      Future.wait.<anonymous closure> (dart:async/future.dart:522:21)
<asynchronous suspension>

[SEVERE] dynamite on lib/src/nextcloud.openapi.json:

Exception: Can not decode "JsonObject" from String
```

Could you try fixing that, so that it will be already compatible? I can't directly tell what's the problem (didn't look into it), but maybe you can.

**Comment by Leptopoda** [comment]
I'm fairly sure it's https://github.com/provokateurin/nextcloud-neon/pull/194/commits/f1188aa941de5037ebda1148bcff0dc2b2e68290

Gonna fix it (probably tomorrow)
Also I thought of making https://github.com/provokateurin/nextcloud-neon/pull/194/commits/756b9e514273b7f87b527e7deef6f5cd2804ebc3 a separate PR although I don't know it's worth it.
I'll need to have a look.

**Comment by Leptopoda** [comment]
@provokateurin are your branches on GH up to date?

I rebased and it works flawlessly on openapi-automatic and openapi-demo fails (although it fails trying to deserialize the schema and not due to the builder)

**Comment by provokateurin** [comment]
Maybe my local branch had some changes, but definitely no big ones. The demo branch can be ignored, since it's broken with the current version of dynamite too.

**Comment by Leptopoda** [comment]
Then maybe your version of the built_value branch was out of date I fixed something similar a few pushes ago

https://github.com/provokateurin/nextcloud-neon/blob/f79797e6d42ceb904b283d0006bfd6c8704fb6a0/packages/dynamite/lib/src/type_result/base.dart#L25-L37

I mean the error must come from here and I definetly added a "JsonObject" case there.

Here you can check out my rebase :)
Didn't bother fixing the one test.
https://github.com/Leptopoda/nextcloud-neon/tree/rebase

**Comment by provokateurin** [comment]
Ah that line of code definitely wasn't there. I think your branch on GH was outdated since I cloned it directly before doing the rebase.

**Comment by provokateurin** [comment]
Next rebase :/ I'm so sorry for not reviewing this yet :sob:

**Comment by provokateurin** [comment]
Sounds good. Could we use that branch to rebuild the output after every change? Without seeing the output it's much harder to review this thing.

**Comment by Leptopoda** [comment]
Yep. I just removed it as it made every rebase a tedious game of rebuilding :)

**Comment by provokateurin** [comment]
Yeah that's what I thougt too. If you do small PRs though, that should work just fine.

**Comment by provokateurin** [comment]
Ok https://github.com/provokateurin/nextcloud-neon/tree/feature/built_value is there now

**Comment by provokateurin** [comment]
Ok so we got almost everything in now, right?

**Comment by Leptopoda** [comment]
we are only missing currently unused parts of dynamite:
- contentString (I'll rewrite and simplify my patch)
- header serialization (I don't think the current implementation is that good but I just ported it as is) the patch should still be ready
- cleanup by removig `extraJsonSerializableValues` as we don't need it anymore and it's not gonna work eith built_value anyways
- migrate neon (the patch should be ready to go)

**Comment by Leptopoda** [comment]
I can rebase this one onto the already merges stuff (or at least push my rebase)

**Comment by provokateurin** [comment]
Sounds good. If you want to fix the header serialization please do, but porting the current implementation is also fine. Rebasing this branch would be nice to see better what is left.

**Comment by Leptopoda** [comment]
Ok so the thing with the content String is that the current patch will generate a new class for every contentString but we could have a general one and use that 
```dart
class ContentString<T>{
  T value;
}

class Test{
  ContentString<List<String>> myParamAsContentString
}
```

We would also not need to generate custom serializers for them as a generic one for the `ContentString` class.

**Comment by provokateurin** [comment]
Great idea. I noticed this too, but didn't bother fixing it because I needed it to just work

**Comment by Leptopoda** [comment]
Ok just changed the ContentString stuff and everything should be ready now.

Do you want one PR with all 4 commits or separate ones?

**Comment by provokateurin** [comment]
Separate where it makes sense please

**Comment by provokateurin** [comment]
I'd suggest after we got all the small PRs in you update this pull request to be the branch all the PRs went into and then we can move from there.

**Comment by Leptopoda** [comment]
other than the rebuilding there were only two conflicts :)

**Comment by Leptopoda** [comment]
Ok I just added a commit dealing with the coverage ignore lines https://github.com/provokateurin/nextcloud-neon/pull/194/commits/e1aadcc493fa9acd897f868e0b1bb43d210ee552

**Comment by provokateurin** [comment]
Did you check the coverage afterwards?

**Comment by Leptopoda** [comment]
Yes but I don't know what it should be :)

It's just adjusting the serialization part (toJson, fromJson) for now.

**Comment by Leptopoda** [comment]
Nothing missing that needs to go into this PR :)

**Comment by Leptopoda** [comment]
Continuation of https://github.com/provokateurin/nextcloud-neon/pull/248#issuecomment-1513054310:

rebasing `feature/spreed` onto this one fails quite spectacularly in multiple cases:
1. `discriminators` are not supported for oneOf or anyOf (while they are working we face a small Type issue)
2. we are again facing the anyOf shall have at least 1 element problem (also on the current version)
3. for some reason https://github.com/provokateurin/nextcloud-neon/blob/1ef9a2ca6ad2da46134bedcba34aa404ff901c28/specs/spreed.json#L523-L557 is completely breaking the build. I think this might be a built_value bug as dynamite runs fine but the built_value generator is somehow failing in ways I haven't seen yet.
4. we'll see :)

At least I now have something to test against with the headers.

**Comment by provokateurin** [comment]
3. I guess the dots in the property names break it

**Comment by provokateurin** [comment]
I'll try rebasing the spreed branch on the latest main soon, but that shouldn't affect any of your work.

**Comment by Leptopoda** [comment]
> 3\. I guess the dots in the property names break it

I'm not that sure but we'll see

> I'll try rebasing the spreed branch on the latest main soon, but that shouldn't affect any of your work.

I quite like having something to test against so having it working is a big win so I just patched it for now.

---

Heres the patch I'll use from now on:
I took inspiration from `feature/openapi-demo` for the unrecognized keys and fixed the anyOf thing.
It's both working on the current `feature/spreed` and `main`.
[0001-fix-spec.patch.txt](https://github.com/provokateurin/nextcloud-neon/files/11265366/0001-fix-spec.patch.txt)

**Comment by provokateurin** [comment]
I'll keep it working. Spreed should be quite easy to migrate because it only needs to be move to neon_spreed without any code refactoring. From you patch I can see you noticed that the branch isn't even compatible with the latest spreed release. I'll apply your patch to get it running at least.

**Comment by provokateurin** [comment]
I rebased it and added your patch, although your patch had one property set to required while it was actually nullable.

**Comment by provokateurin** [comment]
Surprisingly it was still working as good as when I last touched it ~4 months ago :D

**Comment by Leptopoda** [comment]
Ok I think (3) is a bug in the built_value_generator :(
https://github.com/google/built_value.dart/issues/1233

TL;DR:

Let's see. For now we can only escape invalid field names without a `$` sign wich would work for letters by appending a _ (`assert_`, `class_` ...) but does not work on number named fields. For those we need to prefix them with a letter like `n10` or `n20` for now :(

**Comment by provokateurin** [comment]
Ok, I guess we have to go with that if there is no other option. Could we convert the `.` to `Dot`? Otherwise that information gets lost.

**Comment by Leptopoda** [comment]
I don't think dart fields are allowed to start with a number. We'd need something like `n1Dot0` and that's garbage :(

I'll wait for a respond to the issue and see what we'll do going forward.

**Comment by provokateurin** [comment]
Alright, but I think it doesn't need to be resolved for this PR to be merged.

**Comment by Leptopoda** [comment]
I've made a PR against built_value with a fix.
It was fairly trivial :)

For now I'm gonna go forward with using my patched built_value_generator and we can still change our naming if they reject it.

https://github.com/leptopoda/built_value.dart/tree/fix/%231233

**Comment by Leptopoda** [comment]
Ok ContentString and headers will print a warning referencing the issue I just created.
ContentString will still be supported but just as a basic string for now.

Other than that finally ready for merge :tada:

**Comment by provokateurin** [comment]
I'd wait for the built_value_generator.

Can you change the commit messages to have the component they are updating as the prefix? Just look at how I do the commit messages.

> Ok ContentString and headers will print a warning referencing the issue I just created.

So headers are not supported or did you revert back to the old method I built?

**Comment by Leptopoda** [comment]
> Can you change the commit messages to have the component they are updating as the prefix? Just look at how I do the commit messages.

will do that

> So headers are not supported or did you revert back to the old method I built?
I didn't change it as It's not easily possible with built_value. We just can't do the `extraJsonKeyValues` thing.
I could try and fix it but I think it's better to wait that extra week for now :)

**Comment by provokateurin** [comment]
> I'd wait for the built_value_generator.

I'll still review this PR, just wanted to say we should wait for that before merging.

> should the first commit for example be both nextcloud and dynamite?

Could you put all the generated code changes into one commit again? Then you just need to bump the nextcloud package once.

**Comment by Leptopoda** [comment]
I just dropped the built_value_generator patch altogether as we don't hit any limitations for now.

**Comment by Leptopoda** [comment]
I didn't touch headers yet but contentString is finally working and in a somewhat nice/clean way

And here a patch for the spreed tests with built_value
[0001-adapt-spreed-tests-to-built_value.patch.txt](https://github.com/provokateurin/nextcloud-neon/files/11374059/0001-adapt-spreed-tests-to-built_value.patch.txt)

**Comment by provokateurin** [comment]
Awesome! Thanks for all your efforts and patience :heart:

File: packages/dynamite/dynamite/lib/src/openapi_builder.dart [hunk 1]
```diff
@@ -1210,6 +1216,7 @@ class State {
   final resolvedTypes = <String>[];
   final registeredJsonObjects = <String>[];
   final output = <Spec>[];
+  final resolvedTypeCombinations = <TypeResult>[];
```
Comment (Leptopoda) [inline packages/dynamite/dynamite/lib/src/openapi_builder.dart:1219]: Better naming welcome :)

File: packages/.github/workflows/dart.yml [hunk 2]
```diff
@@ -0,0 +1,796 @@
+# Created with package:mono_repo v6.5.0
```
Comment (provokateurin) [inline packages/.github/workflows/dart.yml:1]: This is placed wrongly here

File: packages/dynamite/dynamite/lib/src/type_result/type_result.dart [hunk 3]
```diff
@@ -1,9 +1,22 @@
 part of '../../dynamite.dart';
 
 abstract class TypeResult {
-  TypeResult(this.name);
+  TypeResult(this.className, [this.generics = const []])
```
Comment (provokateurin) [inline packages/dynamite/dynamite/lib/src/type_result/type_result.dart:4]: Could you make generics an optional named parameter? Just to avoid confusion

File: packages/dynamite/dynamite/lib/src/type_result/object.dart [hunk 4]
```diff
@@ -8,14 +8,25 @@ class TypeResultObject extends TypeResult {
         );
 
   @override
-  String serialize(final String object) => '$object.toJson()';
+  String serialize(final String object) {
+    if (className == 'ContentString') {
```
Comment (provokateurin) [inline packages/dynamite/dynamite/lib/src/type_result/object.dart:15]: I think using a constant for this string would better

**Comment by provokateurin** [comment]
Nice work, I left some minor suggestions. Make sure to rebase at some point, as there are currently conflicts on this branch.

File: packages/.github/workflows/dart.yml (hunk 2 above)
Comment (Leptopoda) [reply packages/.github/workflows/dart.yml:1]: oh I thought `mono_repo generate` will also run relative to the project root :eyes:

Comment (provokateurin) [reply packages/.github/workflows/dart.yml:1]: Apparently not, maybe it's a bug? Doesn't seem to be too useful for me to not have it relative to the project

File: packages/dynamite/dynamite/lib/src/type_result/object.dart (hunk 4 above)
Comment (Leptopoda) [reply packages/dynamite/dynamite/lib/src/type_result/object.dart:15]: what do you mean?

**Comment by Leptopoda** [comment]
Yeah I'll probably just do the header stuff and rebase then

File: packages/dynamite/dynamite/lib/src/type_result/object.dart (hunk 4 above)
Comment (provokateurin) [reply packages/dynamite/dynamite/lib/src/type_result/object.dart:15]: Just make a constant variable somewhere containing `'ContentString'` so that you don't have to type it over and over. Might prevent some future typos.

Comment (Leptopoda) [reply packages/dynamite/dynamite/lib/src/type_result/object.dart:15]: Yeah in theory the serialization under the contentstring can be used for all types. So no `TypeResult` would need to override it. I just didn't want to change to much already reviewed code :)

I'll change as suggested and we'll see after this finally lands :+1:

**Comment by Leptopoda** [comment]
With the fix for strings :)

```dart
class _$NextcloudSpreedGetMessagesHeadersSerializer implements StructuredSerializer<NextcloudSpreedGetMessagesHeaders> {
  @override
  final Iterable<Type> types = const [NextcloudSpreedGetMessagesHeaders, _$NextcloudSpreedGetMessagesHeaders];

  @override
  final String wireName = 'NextcloudSpreedGetMessagesHeaders';

  @override
  Iterable<Object?> serialize(
    final Serializers serializers,
    final NextcloudSpreedGetMessagesHeaders object, {
    final FullType specifiedType = FullType.unspecified,
  }) {
    throw UnimplementedError();
  }

  @override
  NextcloudSpreedGetMessagesHeaders deserialize(
    final Serializers serializers,
    final Iterable<Object?> serialized, {
    final FullType specifiedType = FullType.unspecified,
  }) {
    final result = NextcloudSpreedGetMessagesHeadersBuilder();

    final iterator = serialized.iterator;
    while (iterator.moveNext()) {
      final key = iterator.current! as String;
      iterator.moveNext();
      final value = iterator.current as String?;
      switch (key) {
        case 'x-chat-last-given':
          result.xChatLastGiven = value!;
          break;
        case 'x-chat-last-common-read':
          result.xChatLastCommonRead = jsonDecode(value!) as int?;
          break;
      }
    }

    return result.build();
  }
}
```

**Review by provokateurin — CHANGES_REQUESTED** [review]

**Comment by Leptopoda** [comment]
Can you give me an estimate of when we might be able to land this one?

**Comment by provokateurin** [comment]
Today, I hope :)

**Comment by provokateurin** [comment]
LGTM, there is this one commit in the beginning that updates the dependencies, but that is already done in the merged PR, right?

**Review by provokateurin — APPROVED** [review]

**Comment by provokateurin** [comment]
Party time :tada: :tada: :tada:
